File: src/python/algorithms/priorityQueue/taskScheduler/task_scheduler.py
# Task Scheduler
# Given a list of tasks and cooldown n, return the minimum CPU intervals
# needed to finish all tasks (same task must wait n intervals between runs).
# TC: O(T log 26) ~= O(T), SC: O(1)

import heapq

def least_intervals(tasks: list[str], n: int) -> int:
    char_frequency = [0] * 26
    for task in tasks:
        char_frequency[ord(task) - ord('A')] += 1

    # Max-heap via negated values
    max_heap = [-f for f in char_frequency if f > 0]
    heapq.heapify(max_heap)

    time = 0

    while max_heap:
        temp = []
        cycle = n + 1

        while cycle > 0 and max_heap:
            curr = -heapq.heappop(max_heap)
            if curr > 1:
                temp.append(-(curr - 1))
            time += 1
            cycle -= 1

        for val in temp:
            heapq.heappush(max_heap, val)

        if max_heap:
            time += cycle  # idle slots

    return time


# Test cases
task_sets = [
    list("AAABBB"),
    list("AAABBB"),
    list("AAABBBCC"),
    list("AAAAA"),
    list("ABCDEF"),
    list("AABB"),
    list("AAAB"),
    list("AA"),
    list("A"),
    list("AAABBC"),
]
ns       = [2, 0, 2, 2, 2, 1, 2, 3, 0, 2]
expected = [8, 6, 8, 13, 6, 4, 7, 5, 1, 7]

for i in range(len(task_sets)):
    result = least_intervals(task_sets[i], ns[i])
    status = "✓" if result == expected[i] else f"✗ (expected {expected[i]})"
    print(f'tasks="{"".join(task_sets[i])}", n={ns[i]} => {result} {status}')

File: src/javascript/algorithms/priorityQueue/taskScheduler/taskScheduler.js
// Task Scheduler
// Given a list of tasks and cooldown n, return the minimum CPU intervals
// needed to finish all tasks (same task must wait n intervals between runs).
// TC: O(T log 26) ~= O(T), SC: O(1)

class MaxHeap {
  constructor() { this.heap = []; }
  push(val) {
    this.heap.push(val);
    this._bubbleUp(this.heap.length - 1);
  }
  pop() {
    const top = this.heap[0];
    const last = this.heap.pop();
    if (this.heap.length > 0) {
      this.heap[0] = last;
      this._siftDown(0);
    }
    return top;
  }
  isEmpty() { return this.heap.length === 0; }
  _bubbleUp(i) {
    while (i > 0) {
      const parent = Math.floor((i - 1) / 2);
      if (this.heap[parent] >= this.heap[i]) break;
      [this.heap[parent], this.heap[i]] = [this.heap[i], this.heap[parent]];
      i = parent;
    }
  }
  _siftDown(i) {
    const n = this.heap.length;
    while (true) {
      let largest = i;
      const l = 2 * i + 1, r = 2 * i + 2;
      if (l < n && this.heap[l] > this.heap[largest]) largest = l;
      if (r < n && this.heap[r] > this.heap[largest]) largest = r;
      if (largest === i) break;
      [this.heap[largest], this.heap[i]] = [this.heap[i], this.heap[largest]];
      i = largest;
    }
  }
}

function leastIntervals(tasks, n) {
  const charFrequency = new Array(26).fill(0);
  for (const task of tasks) {
    charFrequency[task.charCodeAt(0) - 65]++;
  }

  const maxHeap = new MaxHeap();
  for (const freq of charFrequency) {
    if (freq > 0) maxHeap.push(freq);
  }

  let time = 0;

  while (!maxHeap.isEmpty()) {
    const temp = [];
    let cycle = n + 1;

    while (cycle > 0 && !maxHeap.isEmpty()) {
      const curr = maxHeap.pop();
      if (curr > 1) temp.push(curr - 1);
      time++;
      cycle--;
    }

    for (const val of temp) maxHeap.push(val);

    if (!maxHeap.isEmpty()) {
      time += cycle; // idle slots
    }
  }

  return time;
}

// Test cases
const taskSets = [
  [...'AAABBB'],
  [...'AAABBB'],
  [...'AAABBBCC'],
  [...'AAAAA'],
  [...'ABCDEF'],
  [...'AABB'],
  [...'AAAB'],
  [...'AA'],
  [...'A'],
  [...'AAABBC'],
];
const ns       = [2, 0, 2, 2, 2, 1, 2, 3, 0, 2];
const expected = [8, 6, 8, 13, 6, 4, 7, 5, 1, 7];

for (let i = 0; i < taskSets.length; i++) {
  const result = leastIntervals(taskSets[i], ns[i]);
  const status = result === expected[i] ? "✓" : `✗ (expected ${expected[i]})`;
  console.log(`tasks="${taskSets[i].join('')}", n=${ns[i]} => ${result} ${status}`);
}
